### PR TITLE
Use PngWriter in Texture2D.PlatformSaveAsPng method on Windows, Mac and iOS

### PR DESCRIPTION
**Pull request by disportium**
Not sure whether to place this in the Texture2D.SaveAsPng method or the Texture2D.OpenGL.PlatformSaveAsPng method.  Opted for the latter.

### CODE REVIEW

**Comment by tomspilman** [comment]
@disportium 

So first why for iOS only to begin with?  Just the platform you are testing at the moment?

Also maybe you should start to implement support for multiple source formats like I outlined here:  https://github.com/mono/MonoGame/pull/3537#discussion_r24734078

**Comment by disportium** [comment]
@tomspilman That's right, I'm only testing iOS at the moment, but I have access to Windows and Mac as well, so I'll test on those platforms and modify the compiler directives accordingly.

Will take a look at the multiple source formats issue.

**Comment by disportium** [comment]
Texture surface format support has been added, but only for those surface formats that have classes in the Graphics.PackedVector namespace (Dxt1 and Rg32 don't have such classes, for instance).

PngWriter can also write images with alpha now.

The PlatformSaveAsPng methods of Texture2D.OpenGL and Texture2D.DirectX now use PngWriter for Windows, Mac and iOS.

File: MonoGame.Framework/Graphics/Texture2D.DirectX.cs [hunk 1]
```diff
@@ -268,9 +269,8 @@ private void PlatformSaveAsPng(Stream stream, int width, int height)
 #if WINDOWS_STOREAPP
             SaveAsImage(BitmapEncoder.PngEncoderId, stream, width, height);
```
Comment (tomspilman) [inline MonoGame.Framework/Graphics/Texture2D.DirectX.cs]: If it works on one DX platform I suspect it should work on all.  So you can remove the `#if WINDOWS_STOREAPP` here.

File: MonoGame.Framework/Graphics/Texture2D.OpenGL.cs [hunk 2]
```diff
@@ -489,8 +490,9 @@ private void PlatformSaveAsJpeg(Stream stream, int width, int height)
 
         private void PlatformSaveAsPng(Stream stream, int width, int height)
         {
-#if MONOMAC || WINDOWS
-            SaveAsImage(stream, width, height, ImageFormat.Png);
+#if MONOMAC || WINDOWS || IOS
+            var pngWriter = new PngWriter();
+            pngWriter.Write(this, stream);
 #elif ANDROID
             SaveAsImage(stream, width, height, Bitmap.CompressFormat.Png);
```
Comment (tomspilman) [inline MonoGame.Framework/Graphics/Texture2D.OpenGL.cs]: What is missing for using this on Android?  Testing?

**Comment by tomspilman** [comment]
Looking good!

> Dxt1 and Rg32 don't have such classes

`Rg32` is just an oversight.  I've added an issue for someone to implement it: https://github.com/mono/MonoGame/issues/3605

DXT i'm not sure if even XNA supports saving those.  I suppose we should test and see to know.  MonoGame does have a DXT decompressor here:

https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Graphics/DxtUtil.cs#L48

So it would be possible to support if we wanted.

File: MonoGame.Framework/Utilities/Png/PngWriter.cs [hunk 3]
```diff
@@ -206,5 +203,98 @@ private int CalculateBytesPerPixel()
                     return -1;
             }
         }
+
+        private void GetColorData(Texture2D texture2D)
```
Comment (tomspilman) [inline MonoGame.Framework/Utilities/Png/PngWriter.cs]: Note that we would want to lift this method out into a common location eventually.  So for example we could share it with the JPEG writer too.

File: MonoGame.Framework/Graphics/Texture2D.OpenGL.cs (hunk 2 above)
Comment (disportium) [inline MonoGame.Framework/Graphics/Texture2D.OpenGL.cs]: Yes, in theory it should work. Don't have access to any Android devices
though.

File: MonoGame.Framework/Utilities/Png/PngWriter.cs (hunk 3 above)
Comment (disportium) [inline MonoGame.Framework/Utilities/Png/PngWriter.cs]: Agreed, it's too general to be confined to `PngWriter`.  `Texture2D` seems like a good place for it eventually.

**Comment by disportium** [comment]
Removed WINDOWS_STOREAPP directive in Texture2D.DirectX as requested.

**Comment by tomspilman** [comment]
Thanks.... merging!

**Comment by tomspilman** [comment]
@disportium - FYI.  The remaining packed vector types have been added as of https://github.com/mono/MonoGame/pull/3622.

**Comment by disportium** [comment]
@tomspilman Cool - available in PR #3631

**Pull request merged**